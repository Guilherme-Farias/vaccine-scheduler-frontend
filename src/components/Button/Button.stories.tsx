/* eslint-disable jsx-a11y/control-has-associated-label */
import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import Button from '.';

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: { required: true },
      description: 'Button child',
    },
    variant: {
      description: 'Chose the button variant',
      defaultValue: 'fill',
      control: {
        type: 'select',
      },
    },
    size: {
      description: 'Adjust the size',
      defaultValue: 'medium',
      control: {
        type: 'select',
      },
    },
    fullWidth: {
      description: 'Change the width to occupy 100% of the container',
      control: {
        type: 'boolean',
      },
    },
    disabled: {
      description: 'Change the state of the button',
      control: {
        type: 'boolean',
      },
    },
    color: {
      description: 'Change the color',
      defaultValue: 'primary',
      control: {
        type: 'select',
      },
    },
    fontColor: {
      description: 'Change the color',
      defaultValue: 'white',
      control: {
        type: 'select',
      },
    },
  },
} as unknown as ComponentMeta<typeof Button>;

const Template1: ComponentStory<typeof Button> = args => <Button {...args} />;

const Template2: ComponentStory<typeof Button> = args => (
  <table style={{ fontFamily: 'sans-serif' }} cellSpacing={0} cellPadding={0}>
    <tr>
      <th
        style={{
          textAlign: 'center',
          padding: '16px',
          borderBottom: '1px solid darkgray',
          borderRight: '1px solid darkgray',
        }}
      />
      <th
        style={{
          textAlign: 'center',
          padding: '16px',
          borderBottom: '1px solid darkgray',
        }}
      >
        Fill
      </th>
      <th
        style={{
          textAlign: 'center',
          padding: '16px',
          borderBottom: '1px solid darkgray',
        }}
      >
        Outline
      </th>
      <th
        style={{
          textAlign: 'center',
          padding: '16px',
          borderBottom: '1px solid darkgray',
        }}
      >
        Ghost
      </th>
    </tr>
    <tr>
      <td
        style={{
          textAlign: 'center',
          padding: '16px',
          borderRight: '1px solid darkgray',
        }}
      >
        Small
      </td>
      <td style={{ textAlign: 'center', padding: '16px' }}>
        <Button {...args} size="small" variant="fill">
          Button
        </Button>
      </td>
      <td style={{ textAlign: 'center', padding: '16px' }}>
        <Button {...args} size="small" variant="outline">
          Button
        </Button>
      </td>
      <td style={{ textAlign: 'center', padding: '16px' }}>
        <Button {...args} size="small" variant="ghost">
          Button
        </Button>
      </td>
    </tr>
    <tr>
      <td
        style={{
          textAlign: 'center',
          padding: '16px',
          borderRight: '1px solid darkgray',
        }}
      >
        Medium
      </td>
      <td style={{ textAlign: 'center', padding: '16px' }}>
        <Button {...args} size="medium" variant="fill">
          Button
        </Button>
      </td>
      <td style={{ textAlign: 'center', padding: '16px' }}>
        <Button {...args} {...args} size="medium" variant="outline">
          Button
        </Button>
      </td>
      <td style={{ textAlign: 'center', padding: '16px' }}>
        <Button {...args} size="medium" variant="ghost">
          Button
        </Button>
      </td>
    </tr>
    <tr>
      <td
        style={{
          textAlign: 'center',
          padding: '16px',
          borderRight: '1px solid darkgray',
        }}
      >
        Large
      </td>
      <td style={{ textAlign: 'center', padding: '16px' }}>
        <Button {...args} size="large" variant="fill">
          Button
        </Button>
      </td>
      <td style={{ textAlign: 'center', padding: '16px' }}>
        <Button {...args} size="large" variant="outline">
          Button
        </Button>
      </td>
      <td style={{ textAlign: 'center', padding: '16px' }}>
        <Button {...args} size="large" variant="ghost">
          Button
        </Button>
      </td>
    </tr>
  </table>
);

const Template3: ComponentStory<typeof Button> = args => (
  <div
    style={{ display: 'grid', gap: '10px', gridTemplateColumns: '1fr 1fr 1fr' }}
  >
    <Button {...args} color="primary">
      Primary
    </Button>
    <Button {...args} color="secondary">
      Secondary
    </Button>
    <Button {...args} color="tertiary">
      Tertiary
    </Button>
    <Button {...args} color="green">
      Success
    </Button>
    <Button {...args} color="red">
      Danger
    </Button>
    <Button {...args} color="blue">
      Focus
    </Button>
    <Button {...args} color="yellow">
      Warning
    </Button>
  </div>
);

export const Default = Template1.bind({});
Default.args = {
  children: 'Button',
};

export const TableOfVariants = Template2.bind({});
TableOfVariants.args = {
  children: 'Button',
};

export const TableOfColors = Template3.bind({});
TableOfColors.args = {
  children: 'Button',
};
