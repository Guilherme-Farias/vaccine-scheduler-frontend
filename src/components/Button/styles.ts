import styled, { css, DefaultTheme } from 'styled-components/macro';
import { darken, transparentize } from 'polished';
import { ButtonProps } from '.';

type ButtonBaseProps = Pick<
  ButtonProps,
  'size' | 'fullWidth' | 'color' | 'fontColor'
>;

const buttonBaseModifiers = {
  small: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.md};
    padding: ${theme.spacings['2xs']} ${theme.spacings.sm};
  `,
  medium: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.lg};
    padding: ${theme.spacings.sm} ${theme.spacings.lg};
  `,
  large: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.xl};
    padding: ${theme.spacings.md} ${theme.spacings['3xl']};
  `,
  fullWidth: () => css`
    width: 100%;
  `,
  disabled: () => css`
    &:disabled {
      cursor: not-allowed;
      filter: saturate(30%);
    }
  `,
};

export const ButtonBase = styled.button<ButtonBaseProps>`
  ${({ theme, size, fullWidth, color, disabled }) => css`
    border: 2px solid transparent;
    border-radius: ${theme.border.radius};
    font-weight: ${theme.font.medium};

    ${!!size && buttonBaseModifiers[size](theme)}
    ${!!fullWidth && buttonBaseModifiers.fullWidth()}
    ${!!disabled && buttonBaseModifiers.disabled()}
    &:focus {
      outline-color: ${theme.colors[color!]};
      outline-offset: 4px;
    }
    transition: background-color 0.2s;
  `}
`;

export const FillButton = styled(ButtonBase)`
  ${({ theme, color, fontColor }) => css`
    background-color: ${theme.colors[color!]};
    color: ${theme.colors[fontColor!]};

    &:hover {
      background-color: ${darken(0.05, theme.colors[color!])};
    }
  `}
`;

export const OutlineButton = styled(ButtonBase)`
  ${({ theme, color }) => css`
    background-color: transparent;
    color: ${theme.colors[color!]};
    border: 2px solid currentColor;

    &:hover {
      background-color: ${transparentize(0.9, theme.colors[color!])};
    }
  `}
`;

export const GhostButton = styled(ButtonBase)`
  ${({ theme }) => css`
    background-color: transparent;
    color: ${theme.colors.gray};

    &:focus {
      outline-color: ${theme.colors.gray};
    }
    &:hover {
      background-color: ${transparentize(0.85, theme.colors.gray)};
      color: ${theme.colors.black};
    }
  `}
`;
