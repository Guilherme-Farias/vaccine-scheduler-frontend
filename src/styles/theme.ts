const BREAKPOINTS = {
  phone: 768,
  tablet: 1100,
  laptop: 1500,
};

export default {
  grid: {
    container: '81.25rem', // 1300px
    gutter: '2rem', // 32px
  },
  border: {
    radius: '0.625rem', // 4px
  },
  font: {
    family:
      '"Poppins", -apple-system, BlinkMacSystemFont, "Segoe UI",Roboto, Oxygen, Ubuntu, Cantarell, "Helvetica Neue", sans-serif',
    light: 300,
    regular: 400,
    medium: 500,
    semiBold: 600,
    bold: 700,
    sizes: {
      xs: '0.75rem', // 12px
      sm: '0.875rem', // 14px
      md: '1rem', // 16px
      lg: '1.125rem', // 18px
      xl: '1.25rem', // 20px
      '2xl': '1.5rem', // 24px
      '3xl': '1.875rem', // 30px
      '4xl': '2.25rem', // 36px
      '5xl': '2.5rem', // 40px
      '6xl': '3rem', // 48px
      '7xl': '3.75rem', // 60px
    },
  },
  colors: {
    primary: '#D03E35',
    secondary: '#E37236',
    tertiary: '#FFC801',

    mainBg: '#FFFFFF',
    minorBg: '#D03E35',

    green: '#15803d',
    red: '#DC2626',
    yellow: '#ffc107',
    blue: '#2563eb',

    white: '#FFFFFF',
    lightGray: '#e5e7eb',
    gray: '#6b7280',
    darkGray: '#1f2937',
    black: '#171923',
  },
  layers: {
    base: 100,
    menu: 200,
    overlay: 300,
    modal: 400,
    alwaysOnTop: 500,
  },
  spacings: {
    '3xs': '0.125rem', // 2px
    '2xs': '0.25rem', // 4px
    xs: '0.5rem', // 8px
    sm: '0.75rem', // 12px
    md: '1rem', // 16px
    lg: '1.25rem', // 20px
    xl: '1.5rem', // 24px
    '2xl': '1.75rem', // 28px
    '3xl': '2rem', // 32px
    '4xl': '2.25rem', // 36px
    '5xl': '2.5rem', // 40px
    '6xl': '3rem', // 48px
    '7xl': '3.5rem', // 56px
    '8xl': '4rem', // 64px
    '9xl': '5rem', // 80px
  },
  queries: {
    phoneAndSmaller: `(max-width: ${BREAKPOINTS.phone / 16}rem)`,
    tabletAndSmaller: `(max-width: ${BREAKPOINTS.tablet / 16}rem)`,
    laptopAndSmaller: `(max-width: ${BREAKPOINTS.laptop / 16}rem)`,
  },
} as const;
