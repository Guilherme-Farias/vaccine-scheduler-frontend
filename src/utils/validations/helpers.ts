import Joi from 'joi';
import { IParamErrors } from 'types/global';

export function getFieldErrors(objError: Joi.ValidationResult) {
  const errors: IParamErrors = {};

  if (objError.error) {
    objError.error.details.forEach(err => {
      errors[err.path.join('.')] = err.message;
    });
  }

  return errors;
}
