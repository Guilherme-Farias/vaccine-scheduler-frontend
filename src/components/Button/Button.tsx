import React, { ButtonHTMLAttributes } from 'react';

import theme from 'styles/theme';
import * as S from './styles';

type ThemeColors = typeof theme.colors;

export type ButtonVariants = 'fill' | 'outline' | 'ghost';
export type ButtonSizes = 'small' | 'medium' | 'large';
export type ButtonColors = keyof ThemeColors;
export type ButtonFontColors = keyof ThemeColors;

export type ButtonProps = {
  variant?: ButtonVariants;
  size?: ButtonSizes;
  color?: ButtonColors;
  fontColor?: ButtonFontColors;
  fullWidth?: boolean;
} & ButtonHTMLAttributes<HTMLButtonElement>;

const Button: React.FC<ButtonProps> = ({
  children,
  variant = 'fill',
  size = 'medium',
  color = 'primary',
  fontColor = 'white',
  ...props
}) => {
  let Component;
  if (variant === 'fill') {
    Component = S.FillButton;
  } else if (variant === 'outline') {
    Component = S.OutlineButton;
  } else if (variant === 'ghost') {
    Component = S.GhostButton;
  } else {
    throw new Error(`Unrecognized Button variant: ${variant}`);
  }

  return (
    <Component size={size} color={color} fontColor={fontColor} {...props}>
      {children}
    </Component>
  );
};
export default Button;
