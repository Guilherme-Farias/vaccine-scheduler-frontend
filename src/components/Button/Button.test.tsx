import { render, screen } from 'utils/tests/helpers';
import theme from 'styles/theme';
import Button, { ButtonVariants } from '.';

describe('<Button />', () => {
  it('should render the medium size by default', () => {
    const { container } = render(<Button>Test</Button>);
    expect(screen.getByRole('button', { name: /test/i })).toHaveStyle({
      'font-size': '1.125rem',
      padding: '0.75rem 1.25rem',
    });

    expect(container.firstChild).toMatchSnapshot();
  });
  it('should render the small size', () => {
    render(<Button size="small">Test</Button>);
    expect(screen.getByRole('button', { name: /test/i })).toHaveStyle({
      'font-size': '1rem',
      padding: '0.25rem 0.75rem',
    });
  });
  it('should render the large size', () => {
    render(<Button size="large">Test</Button>);
    expect(screen.getByRole('button', { name: /test/i })).toHaveStyle({
      'font-size': '1.25rem',
      padding: '1rem 2rem',
    });
  });
  it('should render a fullWidth version', () => {
    render(<Button fullWidth>Test</Button>);
    expect(screen.getByRole('button', { name: /test/i })).toHaveStyle({
      width: '100%',
    });
  });
  it('should render the fill variant by default', () => {
    render(<Button>Test</Button>);
    expect(screen.getByRole('button', { name: /test/i })).toHaveStyle({
      'background-color': theme.colors.primary,
      color: theme.colors.white,
    });
  });
  it('should render the outline variant', () => {
    render(<Button variant="outline">Test</Button>);
    expect(screen.getByRole('button', { name: /test/i })).toHaveStyle({
      'background-color': 'transparent',
      color: theme.colors.primary,
      border: '2px solid currentColor',
    });
  });
  it('should render the ghost variant', () => {
    render(<Button variant="ghost">Test</Button>);
    expect(screen.getByRole('button', { name: /test/i })).toHaveStyle({
      'background-color': 'transparent',
      color: theme.colors.gray,
    });
  });
  it('should throw error when unrecognized Button variant is provided', () => {
    // Restore writing to stderr.
    const err = console.error;
    console.error = jest.fn();

    const unrecognized = 'unrecognized variant' as unknown as ButtonVariants;
    expect(() =>
      render(<Button variant={unrecognized}>Test</Button>),
    ).toThrow();

    // Restore writing to stderr.
    console.error = err;
  });

  it('should be able to change color and font color', () => {
    render(
      <Button color="tertiary" fontColor="black">
        Test
      </Button>,
    );
    expect(screen.getByRole('button', { name: /test/i })).toHaveStyle({
      'background-color': theme.colors.tertiary,
      color: theme.colors.black,
    });
  });

  it('should render a disabled Button', () => {
    render(<Button disabled>Test</Button>);

    expect(screen.getByRole('button', { name: /test/i })).toHaveStyleRule(
      'cursor',
      'not-allowed',
      {
        modifier: ':disabled',
      },
    );
  });
});
