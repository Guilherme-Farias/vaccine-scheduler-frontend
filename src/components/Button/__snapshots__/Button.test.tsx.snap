// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size by default 1`] = `
.c0 {
  border: 2px solid transparent;
  border-radius: 0.625rem;
  font-weight: 500;
  font-size: 1.125rem;
  padding: 0.75rem 1.25rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
}

.c0:focus {
  outline-color: #D03E35;
  outline-offset: 4px;
}

.c1 {
  background-color: #D03E35;
  color: #FFFFFF;
}

.c1:hover {
  background-color: #bf352c;
}

<button
  class="c0 c1"
  color="primary"
>
  Test
</button>
`;
